### PR TITLE
Convert nanopb struct to GDTCOREventDataObject and update the source code to use nanopb struct.

### PR DESCRIPTION
**Pull request by jeremyjiang-dev**
## How to review this giant PR?
1. The key changes is about converting nanopb struct object to `GDTCOREventDataObject` so that we can dispatch the data. See `FPRClient.m  transportBytes()`.
2. `FPR` prefixes are added back to `FPRNanoPbUtils.h` and `FPRNanoPbUtils.m`. `FPRProtoUtils.h` and `FPRProtoUtils.m` are commented out to prevent symbol collision .
3. A few methods are added to help set fields of perf metric struct object in `FPRNanoPbUtils.h` and `FPRNanoPbUtils.m`.
4. The rest of the changes are about replacing proto object with nanopb struct object.

### CODE REVIEW

File: FirebasePerformance/Sources/FPRClient.m [hunk 1]
```diff
@@ -127,7 +125,7 @@ - (BOOL)startWithConfiguration:(FPRConfiguration *)config error:(NSError *__auto
 
 #ifdef TARGET_HAS_MOBILE_CONNECTIVITY
     // Create telephony network information object ahead of time to avoid runtime delays.
-    FPRNetworkInfo();
+    NetworkInfo();
```
Comment (paulb777) [inline FirebasePerformance/Sources/FPRClient.m:128]: The prefix should stay on this symbol.  Otherwise there is the potential for build failures for any client apps that have a `NetworkInfo` function in their app or other library.

File: FirebasePerformance/Sources/FPRClient.m [hunk 2]
```diff
@@ -236,20 +237,20 @@ - (void)logGaugeMetric:(nonnull NSArray *)gaugeData forSessionId:(nonnull NSStri
     return;
   }
   dispatch_group_async(self.eventsQueueGroup, self.eventsQueue, ^{
-    FPRMSGPerfMetric *metric = FPRGetPerfMetricMessage(self.config.appID);
-    FPRMSGGaugeMetric *gaugeMetric = nil;
+    firebase_perf_v1_PerfMetric metric = GetPerfMetricMessage(self.config.appID);
+    firebase_perf_v1_GaugeMetric gaugeMetric = firebase_perf_v1_GaugeMetric_init_default;
     if ((gaugeData != nil && gaugeData.count != 0) && (sessionId != nil && sessionId.length != 0)) {
-      gaugeMetric = FPRGetGaugeMetric(gaugeData, sessionId);
+      gaugeMetric = GetGaugeMetric(gaugeData, sessionId);
```
Comment (paulb777) [inline FirebasePerformance/Sources/FPRClient.m:243]: Restore prefix

File: FirebasePerformance/Sources/FPRNanoPbUtils.h [hunk 3]
```diff
@@ -136,6 +136,58 @@ extern firebase_perf_v1_GaugeMetric GetGaugeMetric(NSArray* _Nonnull gaugeData,
  */
 extern firebase_perf_v1_ApplicationProcessState ApplicationProcessState(FPRTraceState state);
 
+/** Populate a firebase_perf_v1_PerfMetric object with the given firebase_perf_v1_ApplicationInfo.
```
Comment (paulb777) [inline FirebasePerformance/Sources/FPRNanoPbUtils.h:139]: C Function externs should be prefixed.

**Comment by paulb777** [comment]
The installations-test failure is not related to this PR.

**Comment by jeremyjiang-dev** [comment]
> The installations-test failure is not related to this PR.

Thanks Paul! Do you have any clue why `test_coverage / create_report (pull_request)` failed?

**Comment by paulb777** [comment]
I don't. @granluo would you take a look.  Probably not a blocker though.

File: FirebasePerformance/Sources/FPRClient.m (hunk 1 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRClient.m]: Thanks Paul! I delete it due to symbol collision in FPRProtoUtils files. I have commented FPRProtoUtils files and restored the prefix.

File: FirebasePerformance/Sources/FPRClient.m (hunk 2 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRClient.m]: Updated!

File: FirebasePerformance/Sources/FPRNanoPbUtils.h (hunk 3 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRNanoPbUtils.h:139]: Done.

File: FirebasePerformance/Sources/FPRProtoUtils.h [hunk 4]
```diff
@@ -1,65 +1,65 @@
-// Copyright 2020 Google LLC
+//// Copyright 2020 Google LLC
```
Comment (paulb777) [inline FirebasePerformance/Sources/FPRProtoUtils.h:1]: 2021 - copyrights should have the year they're first published externally

**Review by paulb777 — COMMENTED** [review]
Just a nit about a copyright and LGTM

Deferring approval to @visumickey

**Comment by paulb777** [comment]
There does seem to be a problem with coverage generation:

`2021-07-28 17:46:20.190 xccov[4010:16082] Requested but did not find extension point with identifier `

@granluo

File: FirebasePerformance/Sources/FPRProtoUtils.h (hunk 4 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRProtoUtils.h:1]: FPRProtoUtils.h was first published externally in 2020. I will remove this file and all proto related files in the next PR.

File: FirebasePerformance/Sources/Loggers/FPRGDTEvent.h [hunk 5]
```diff
@@ -15,24 +15,23 @@
...
-@property(nonatomic, readonly) FPRMSGPerfMetric *metric;
+/** firebase_perf_v1_PerfMetric that is going to be converted. */
+@property(nonatomic, readonly) firebase_perf_v1_PerfMetric metric;
 
 - (instancetype)init NS_UNAVAILABLE;
 
-/** Converts a PerfMetric event to a GDTEvent. */
-+ (instancetype)gdtEventForPerfMetric:(FPRMSGPerfMetric *)perfMetric;
+/** Converts a firebase_perf_v1_PerfMetric object to a GDTEvent. */
++ (instancetype)gdtEventForPerfMetric:(firebase_perf_v1_PerfMetric)perfMetric;
```
Comment (visumickey) [inline FirebasePerformance/Sources/Loggers/FPRGDTEvent.h:34]: Curious, NIT: Is there a way to replace `firebase_perf_v1_PerfMetric` to something more relatable. Eg. FPRPerfMetric? The name is too long to perceive :)

File: FirebasePerformance/Sources/Loggers/FPRGDTEvent.m [hunk 6]
```diff
@@ -49,7 +54,30 @@ - (instancetype)initForPerfMetric:(FPRMSGPerfMetric *)perfMetric {
 #pragma mark - GDTCOREventDataObject protocol methods
 
 - (NSData *)transportBytes {
-  return [self.metric data];
+  pb_ostream_t sizestream = PB_OSTREAM_SIZING;
+
+  // Encode 1 time to determine the size.
```
Comment (visumickey) [inline FirebasePerformance/Sources/Loggers/FPRGDTEvent.m:59]: Since this is just needed once, should we make this static and measured just once?

File: FirebasePerformance/Sources/Loggers/FPRGDTLogger.m [hunk 7]
```diff
@@ -54,7 +54,7 @@ - (instancetype)initWithLogSource:(NSInteger)logSource {
   return self;
 }
 
-- (void)logEvent:(FPRMSGPerfMetric *)event {
+- (void)logEvent:(struct _firebase_perf_v1_PerfMetric)event {
```
Comment (visumickey) [inline FirebasePerformance/Sources/Loggers/FPRGDTLogger.m:57]: Why do we need `_` in front of `_firebase_perf_v1_PerfMetric`. Can we use the struct definition that was generated?

File: FirebasePerformance/Tests/Unit/Instruments/FIRHTTPMetricTests.m [hunk 8]
```diff
@@ -187,7 +185,6 @@ - (void)testMetricCreationFailsWhenResponseCodeNotSet {
   [metric markResponseStart];
   [metric stop];
   FPRNetworkTrace *networkTrace = metric.networkTrace;
-  XCTAssertNil(FPRGetNetworkRequestMetric(networkTrace));
```
Comment (visumickey) [inline FirebasePerformance/Tests/Unit/Instruments/FIRHTTPMetricTests.m:190]: Why are these removed? These are valid tests, isn't it?

**Comment by jeremyjiang-dev** [comment]
> There does seem to be a problem with coverage generation:
> 
> `2021-07-28 17:46:20.190 xccov[4010:16082] Requested but did not find extension point with identifier `
> 
> @granluo

@granluo @paulb777  Any updates? Is it safe to merge with this error?

File: FirebasePerformance/Sources/FPRClient.m [hunk 9]
```diff
@@ -271,14 +271,15 @@ - (void)processAndLogEvent:(FPRMSGPerfMetric *)event {
   });
 
   // Attempts to dispatch events if successfully retrieve installation ID.
+  __block firebase_perf_v1_PerfMetric curEvent = event;
```
[the remark this answers is not in the document]
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRClient.m:274]: If I just do `event.application_info.app_instance_id = FPREncodeString(identifier);`, then there is a compiler error `Read-only variable is not assignable`. The reasons is that blocks cannot alter variables outside of the block without that variable having a `__block`.

**Comment by paulb777** [comment]
OK to merge. If the `xccov` failure persists in subsequent PR's, an issue should be opened to track it.

File: FirebasePerformance/Sources/Loggers/FPRGDTEvent.m (hunk 6 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/Loggers/FPRGDTEvent.m:59]: Could you elaborate on this? I am not so clear. `sizestream` is a local variable, do you want to make it static? `transportBytes` is a instance method, we cannot make `sizestream` static.

File: FirebasePerformance/Sources/Loggers/FPRGDTLogger.m (hunk 7 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/Loggers/FPRGDTLogger.m:57]: They are the same. I have changed all of them to `firebase_perf_v1_PerfMetric`.

File: FirebasePerformance/Sources/Loggers/FPRGDTEvent.h (hunk 5 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/Loggers/FPRGDTEvent.h:34]: This may result in a huge change throughout the SDK. I would prefer to make this change in a later PR.

File: FirebasePerformance/Sources/FPRNanoPbUtils.h [hunk 10]
```diff
@@ -107,38 +107,90 @@ extern StringToNumberMap* _Nullable FPREncodeStringToNumberMap(NSDictionary* _Nu
...
+extern firebase_perf_v1_ApplicationProcessState FPRApplicationProcessState(FPRTraceState state);
+
+/** Populate a firebase_perf_v1_PerfMetric object with the given firebase_perf_v1_ApplicationInfo.
+ *
+ *  @param perfMetric The firebase_perf_v1_PerfMetric to be populated.
+ *  @param appInfo The firebase_perf_v1_ApplicationInfo object that will be added to
+ * firebase_perf_v1_PerfMetric.
+ *  @return A firebase_perf_v1_PerfMetric object.
+ */
+extern firebase_perf_v1_PerfMetric FPRSetApplicationInfo(firebase_perf_v1_PerfMetric perfMetric,
```
[the remark this answers is not in the document]
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRNanoPbUtils.h:146]: Sounds good! Done!

> OK to merge. If the `xccov` failure persists in subsequent PR's, an issue should be opened to track it.

Sounds good! Thanks Paul!

File: FirebasePerformance/Tests/Unit/Instruments/FIRHTTPMetricTests.m (hunk 8 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Tests/Unit/Instruments/FIRHTTPMetricTests.m:190]: Because it returns a struct instead of a pointer now. I updated the test to check if the networkTrace is valid. Please take a look.

File: FirebasePerformance/Sources/FPRNanoPbUtils.m [hunk 11]
```diff
@@ -500,3 +503,40 @@ firebase_perf_v1_ApplicationProcessState ApplicationProcessState(FPRTraceState s
 
   return [orderedSessions copy];
 }
+
+#pragma mark - Nanopb struct fields populating helper methods
+
+firebase_perf_v1_PerfMetric FPRSetApplicationInfo(firebase_perf_v1_PerfMetric perfMetric,
```
[the remark this answers is not in the document]
Comment (jeremyjiang-dev) [reply FirebasePerformance/Sources/FPRNanoPbUtils.m:509]: Sounds good! Done!

File: FirebasePerformance/Sources/FPRClient.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (visumickey) [reply FirebasePerformance/Sources/FPRClient.m:274]: Sounds good!

**Review by visumickey — APPROVED** [review]
